### ISSUE TITLE
Error: k2 compiled without CUDA support

### ISSUE THREAD

**Issue by bsshruthi22** [k2-fsa/icefall#836]
I have used the below steps to install k2
conda create --name k2 python=3.8
conda activate k2
pip install git+https://github.com/lhotse-speech/lhotse
conda install -c k2-fsa -c pytorch -c nvidia k2 pytorch=1.13.1 pytorch-cuda=11.7 python=3.8
git clone https://github.com/k2-fsa/icefall
cd icefall/
pip install -r requirements.txt
export PYTHONPATH=/media/administrator/icefall:$PYTHONPATH
cd egs/yesno/ASR
./prepare.sh 
./tdnn/train.py -- ERROR encountered here

**ERROR is as below:**
./tdnn/train.py 
2023-01-11 17:10:56,413 INFO [train.py:481] Training started
2023-01-11 17:10:56,413 INFO [train.py:482] {'exp_dir': PosixPath('tdnn/exp'), 'lang_dir': PosixPath('data/lang_phone'), 'lr': 0.01, 'feature_dim': 23, 'weight_decay': 1e-06, 'start_epoch': 0, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 10, 'reset_interval': 20, 'valid_interval': 10, 'beam_size': 10, 'reduction': 'sum', 'use_double_scores': True, 'world_size': 1, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 15, 'seed': 42, 'feature_dir': PosixPath('data/fbank'), 'max_duration': 30.0, 'bucketing_sampler': False, 'num_buckets': 10, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': False, 'return_cuts': True, 'num_workers': 2, 'env_info': {'k2-version': '1.23.3', 'k2-build-type': 'Release', 'k2-with-cuda': False, 'k2-git-sha1': 'fd609c0ecd074a2414cbfe7cc827427e6b04c8fc', 'k2-git-date': 'Thu Jan 5 04:50:13 2023', 'lhotse-version': '1.12.0.dev+git.3ccfeb7.clean', 'torch-version': '1.13.1+cu117', 'torch-cuda-available': True, 'torch-cuda-version': '11.7', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '42cc101-clean', 'icefall-git-date': 'Mon Jan 9 12:38:39 2023', 'icefall-path': '/media/administrator/hdd/Sishir/icefall', 'k2-path': '/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/__init__.py', 'lhotse-path': '/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/lhotse/__init__.py', 'hostname': 'gpuserver', 'IP address': '10.101.1.248'}}
2023-01-11 17:10:56,414 INFO [lexicon.py:168] Loading pre-compiled data/lang_phone/Linv.pt
2023-01-11 17:10:56,416 INFO [train.py:495] device: cuda:0
[F] /usr/share/miniconda/envs/k2/conda-bld/k2_1672899624360/work/k2/csrc/device_guard.h:66:static int32_t k2::DeviceGuard::GetDevice() k2 compiled without CUDA support

[ Stack-Trace: ]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/lib/libk2_log.so(k2::internal::GetStackTrace()+0x47) [0x7fca3cd3e187]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x78f8a) [0x7fca3d814f8a]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x4b5b9) [0x7fca3d7e75b9]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x218c5) [0x7fca3d7bd8c5]
python3(PyCFunction_Call+0x52) [0x4e0212]
python3(_PyObject_MakeTpCall+0x3eb) [0x4d0eab]
python3() [0x4e897b]
python3(_PyEval_EvalFrameDefault+0x4d38) [0x4cbd18]
python3() [0x4e8717]
python3(_PyEval_EvalFrameDefault+0x4d38) [0x4cbd18]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(_PyFunction_Vectorcall+0x19c) [0x4da1cc]
python3(_PyObject_FastCallDict+0x25f) [0x4d04df]
python3() [0x4e44aa]
python3(_PyObject_MakeTpCall+0x34b) [0x4d0e0b]
python3(_PyEval_EvalFrameDefault+0x5605) [0x4cc5e5]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(_PyFunction_Vectorcall+0x19c) [0x4da1cc]
python3(_PyEval_EvalFrameDefault+0x17bf) [0x4c879f]
python3(_PyFunction_Vectorcall+0x106) [0x4da136]
python3(_PyEval_EvalFrameDefault+0x907) [0x4c78e7]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(PyEval_EvalCodeEx+0x39) [0x4c5c99]
python3(PyEval_EvalCode+0x1b) [0x5691ab]
python3() [0x588a71]
python3() [0x58268f]
python3() [0x58bf82]
python3(PyRun_SimpleFileExFlags+0x1a2) [0x58afa2]
python3(Py_RunMain+0x35c) [0x57ff1c]
python3(Py_BytesMain+0x39) [0x55ced9]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fcacce88083]
python3() [0x55cd8e]

Traceback (most recent call last):
  File "./tdnn/train.py", line 575, in <module>
    main()
  File "./tdnn/train.py", line 571, in main
    run(rank=0, world_size=1, args=args)
  File "./tdnn/train.py", line 497, in run
    graph_compiler = CtcTrainingGraphCompiler(lexicon=lexicon, device=device)
  File "/media/administrator/hdd/Sishir/icefall/icefall/graph_compiler.py", line 51, in __init__
    L_inv = lexicon.L_inv.to(device)
  File "/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/fsa.py", line 1106, in to
    ans = Fsa(self.arcs.to(device), properties=self.properties)
RuntimeError: 
    Some bad things happened. Please read the above error messages and stack
    trace. If you are using Python, the following command may be helpful:

      gdb --args python /path/to/your/code.py

    (You can use `gdb` to debug the code. Please consider compiling
    a debug version of k2.).

    If you are unable to fix it, please open an issue at:

      https://github.com/k2-fsa/k2/issues/new
    
(k2) administrator@gpuserver:/media/administrator/hdd/Sishir/icefall/egs/yesno/ASR$ 
(k2) administrator@gpuserver:/media/administrator/hdd/Sishir/icefall/egs/yesno/ASR$ export CUDA_VISIBLE_DEVICES=0
(k2) administrator@gpuserver:/media/administrator/hdd/Sishir/icefall/egs/yesno/ASR$ ./tdnn/train.py 
2023-01-11 17:12:42,333 INFO [train.py:481] Training started
2023-01-11 17:12:42,333 INFO [train.py:482] {'exp_dir': PosixPath('tdnn/exp'), 'lang_dir': PosixPath('data/lang_phone'), 'lr': 0.01, 'feature_dim': 23, 'weight_decay': 1e-06, 'start_epoch': 0, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 10, 'reset_interval': 20, 'valid_interval': 10, 'beam_size': 10, 'reduction': 'sum', 'use_double_scores': True, 'world_size': 1, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 15, 'seed': 42, 'feature_dir': PosixPath('data/fbank'), 'max_duration': 30.0, 'bucketing_sampler': False, 'num_buckets': 10, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': False, 'return_cuts': True, 'num_workers': 2, 'env_info': {'k2-version': '1.23.3', 'k2-build-type': 'Release', 'k2-with-cuda': False, 'k2-git-sha1': 'fd609c0ecd074a2414cbfe7cc827427e6b04c8fc', 'k2-git-date': 'Thu Jan 5 04:50:13 2023', 'lhotse-version': '1.12.0.dev+git.3ccfeb7.clean', 'torch-version': '1.13.1+cu117', 'torch-cuda-available': True, 'torch-cuda-version': '11.7', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '42cc101-clean', 'icefall-git-date': 'Mon Jan 9 12:38:39 2023', 'icefall-path': '/media/administrator/hdd/Sishir/icefall', 'k2-path': '/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/__init__.py', 'lhotse-path': '/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/lhotse/__init__.py', 'hostname': 'gpuserver', 'IP address': '10.101.1.248'}}
2023-01-11 17:12:42,335 INFO [lexicon.py:168] Loading pre-compiled data/lang_phone/Linv.pt
2023-01-11 17:12:42,336 INFO [train.py:495] device: cuda:0
[F] /usr/share/miniconda/envs/k2/conda-bld/k2_1672899624360/work/k2/csrc/device_guard.h:66:static int32_t k2::DeviceGuard::GetDevice() k2 compiled without CUDA support

[ Stack-Trace: ]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/lib/libk2_log.so(k2::internal::GetStackTrace()+0x47) [0x7fa46f5d5187]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x78f8a) [0x7fa4700abf8a]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x4b5b9) [0x7fa47007e5b9]
/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so(+0x218c5) [0x7fa4700548c5]
python3(PyCFunction_Call+0x52) [0x4e0212]
python3(_PyObject_MakeTpCall+0x3eb) [0x4d0eab]
python3() [0x4e897b]
python3(_PyEval_EvalFrameDefault+0x4d38) [0x4cbd18]
python3() [0x4e8717]
python3(_PyEval_EvalFrameDefault+0x4d38) [0x4cbd18]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(_PyFunction_Vectorcall+0x19c) [0x4da1cc]
python3(_PyObject_FastCallDict+0x25f) [0x4d04df]
python3() [0x4e44aa]
python3(_PyObject_MakeTpCall+0x34b) [0x4d0e0b]
python3(_PyEval_EvalFrameDefault+0x5605) [0x4cc5e5]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(_PyFunction_Vectorcall+0x19c) [0x4da1cc]
python3(_PyEval_EvalFrameDefault+0x17bf) [0x4c879f]
python3(_PyFunction_Vectorcall+0x106) [0x4da136]
python3(_PyEval_EvalFrameDefault+0x907) [0x4c78e7]
python3(_PyEval_EvalCodeWithName+0x1f5) [0x4c5e95]
python3(PyEval_EvalCodeEx+0x39) [0x4c5c99]
python3(PyEval_EvalCode+0x1b) [0x5691ab]
python3() [0x588a71]
python3() [0x58268f]
python3() [0x58bf82]
python3(PyRun_SimpleFileExFlags+0x1a2) [0x58afa2]
python3(Py_RunMain+0x35c) [0x57ff1c]
python3(Py_BytesMain+0x39) [0x55ced9]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fa4ff71f083]
python3() [0x55cd8e]

Traceback (most recent call last):
  File "./tdnn/train.py", line 575, in <module>
    main()
  File "./tdnn/train.py", line 571, in main
    run(rank=0, world_size=1, args=args)
  File "./tdnn/train.py", line 497, in run
    graph_compiler = CtcTrainingGraphCompiler(lexicon=lexicon, device=device)
  File "/media/administrator/hdd/Sishir/icefall/icefall/graph_compiler.py", line 51, in __init__
    L_inv = lexicon.L_inv.to(device)
  File "/media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/fsa.py", line 1106, in to
    ans = Fsa(self.arcs.to(device), properties=self.properties)
RuntimeError: 
    Some bad things happened. Please read the above error messages and stack
    trace. If you are using Python, the following command may be helpful:

      gdb --args python /path/to/your/code.py

    (You can use `gdb` to debug the code. Please consider compiling
    a debug version of k2.).

    If you are unable to fix it, please open an issue at:

      https://github.com/k2-fsa/k2/issues/new
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
Extra info:::
(k2) administrator@gpuserver:$ python3 -m k2.version

Collecting environment information...

k2 version: 1.23.3
Build type: Release
Git SHA1: fd609c0ecd074a2414cbfe7cc827427e6b04c8fc
Git date: Thu Jan 5 04:50:13 2023
Cuda used to build k2: 
cuDNN used to build k2: 
Python version used to build k2: 3.8
OS used to build k2: Ubuntu 18.04.6 LTS
CMake version: 3.18.2
GCC version: 7.5.0
CMAKE_CUDA_FLAGS: 
CMAKE_CXX_FLAGS:  -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-unused-variable  -Wno-strict-overflow 
PyTorch version used to build k2: 1.13.1
PyTorch is using Cuda: 
NVTX enabled: False
With CUDA: False
Disable debug: True
Sync kernels : False
Disable checks: False
Max cpu memory allocate: 214748364800 bytes (or 200.0 GB)
k2 abort: False
__file__: /media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/k2/version/version.py
_k2.__file__: /media/administrator/hdd/anaconda3/envs/k2/lib/python3.8/site-packages/_k2.cpython-38-x86_64-linux-gnu.so
######################################################################################################
python3 -m torch.utils.collect_env

Collecting environment information...
PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-135-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro RTX 4000
Nvidia driver version: 470.161.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cpuonly                   2.0                           0    pytorch
[conda] k2                        1.23.3.dev20230105 cpu_py3.8_torch1.13.1  [cpuonly]  k2-fsa
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.13.1              py3.8_cpu_0    pytorch
[conda] pytorch-cuda              11.7                 h67b0de4_1    pytorch
[conda] pytorch-mutex             1.0                         cpu    pytorch

**Comment by csukuangfj** [comment]
> [F] /usr/share/miniconda/envs/k2/conda-bld/k2_1672899624360/work/k2/csrc/device_guard.h:66:static int32_t k2::DeviceGuard::GetDevice() k2 compiled without CUDA support

> [conda] k2 1.23.3.dev20230105 cpu_py3.8_torch1.13.1 [cpuonly] k2-fsa

You have installed a CPU version of k2.

----
> [pip3] torch==1.13.1
> [conda] pytorch 1.13.1 py3.8_cpu_0 pytorch

Also, it looks like you have used both `pip install` and `conda install` to install PyTorch.

Please don't do that.

----

The fix is to install a CUDA version of k2.

---

As you are using
```
conda install -c k2-fsa -c pytorch -c nvidia k2 pytorch=1.13.1 pytorch-cuda=11.7 python=3.8
```
so a I assume you want to install pytorch 1.13.1 with cuda 11.7.

You can use the following steps to install k2.

(1) Please create a new virtual environment
(2) Install PyTorch 1.13.1
```
wget https://download.pytorch.org/whl/cu117/torch-1.13.1%2Bcu117-cp38-cp38-linux_x86_64.whl
pip install ./torch-1.13.1%2Bcu117-cp38-cp38-linux_x86_64.whl
```
(3) Install cudatoolkit 11.7
Please follow https://k2-fsa.github.io/k2/installation/cuda-cudnn.html#cuda-11-7
(4) Install k2
Please follow https://k2-fsa.github.io/k2/installation/from_source.html

Basically, all you need is
```bash
git clone https://github.com/k2-fsa/k2.git
cd k2
export K2_MAKE_ARGS="-j6"
python3 setup.py install
```
